src/blockmap.jl

```diff
@@ -0,0 +1,477 @@
+    cumulsizes(block_sizes, i, j+1) - cumulsizes(block_sizes, i, j)
+
+function blocksize(block_sizes::AbstractBlockSizes, i::Tuple{Integer, Integer})
+    return blocksize(block_sizes, 1, i[1]), blocksize(block_sizes, 1, i[1])
+end
+
+# Gives the total sizes
+function Base.size(block_sizes::AbstractBlockSizes)
+    @inbounds return cumulsizes(block_sizes, 1)[end] - 1, cumulsizes(block_sizes, 2)[end] - 1
+end
+
+function Base.show(io::IO, block_sizes::AbstractBlockSizes)
+    print(io, " × ", diff(cumulsizes(block_sizes, 2)))
+end
+
+@inline function searchlinear(vec::AbstractVector, a)
+    l = length(vec)
+    @inbounds for i in 1:l
+        vec[i] > a && return i - 1
+    end
+    return l
+end
+
+_find_block(bs::AbstractVector, i::Integer) = length(bs) > 10 ? last(searchsorted(bs, i)) : searchlinear(bs, i)
+
+@inline function _find_block(block_sizes::AbstractBlockSizes, dim::Integer, i::Integer)
+    bs = cumulsizes(block_sizes, dim)
+    block = _find_block(bs, i)
+    @inbounds cum_size = cumulsizes(block_sizes, dim, block) - 1
+    return block, i - cum_size
+end
+
+nblocks(block_sizes::AbstractBlockSizes) = nblocks(block_sizes, 1), nblocks(block_sizes, 2)
+
+# @inline nblocks(block_array::AbstractArray) = nblocks(blocksizes(block_array))
+
+@inline Base.@propagate_inbounds nblocks(block_sizes::AbstractBlockSizes, i::Integer) =
+    length(cumulsizes(block_sizes, i)) - 1
+
+function nblocks(block_sizes::AbstractBlockSizes, i::Integer, j::Integer)
+    b = nblocks(block_sizes)
+    return b[i], b[j]
+end
+
+function Base.copy(block_sizes::BlockSizes)
+    return BlockSizes(copy(cumulsizes(block_sizes, 1)), copy(cumulsizes(block_sizes, 2)))
+end
+
+@inline function globalrange(block_sizes::AbstractBlockSizes, block_index::NTuple{2, Integer})
+    @inbounds v = (cumulsizes(block_sizes, 1, block_index[1]):cumulsizes(block_sizes, 1, block_index[1] + 1) - 1,
+                   cumulsizes(block_sizes, 2, block_index[2]):cumulsizes(block_sizes, 2, block_index[2] + 1) - 1)
+    return v
+end
+
+"""
+    blocksize(A, inds)
+
+Returns a tuple containing the size of the block at block index `inds`.
+"""
+
+@inline blocksize(block_array::AbstractArray, i::Integer...) =
+    blocksize(blocksizes(block_array), i...)
+
+@inline blocksize(block_array::AbstractMatrix{T}, i::Tuple{Integer, Integer}) where {T} =
+    blocksize(blocksizes(block_array), i)
+
+cumulsizes(A::AbstractArray) = cumulsizes(blocksizes(A))
+@inline cumulsizes(A::AbstractArray, i) = cumulsizes(blocksizes(A), i)
+@inline cumulsizes(A::AbstractArray, i, j) = cumulsizes(blocksizes(A), i, j)
+
+##############
+# BlockMaps #
+##############
+
+abstract type AbstractBlockMap{T,As,Bs} <: LinearMap{T} end
+
+struct HBlockMap{T,As<:Tuple{Vararg{LinearMap}},BS<:AbstractBlockSizes} <: AbstractBlockMap{T,As,BS}
+    maps::As
+    block_sizes::BS
+    global function _HBlockMap(maps::R, block_sizes::BS) where {T, R<:Tuple{Vararg{LinearMap{T}}}, BS<:AbstractBlockSizes}
```

Comment:
I guess in the other package they actively want to prevent users to accidentally call the constructor directly, that's why they "divert" the constructors. We don't export any of these types or constructors, so I have simplified this approach.